Improve and supplement LoRA tuning documentation

caozuoba opened the pull request:
<!-- markdownlint-disable -->

## Purpose
Improve and supplement LoRA tuning documentation
## Test Plan

## Test Result
<img width="416" height="158" alt="bc182bf14be8ee10687ffecfe0dda875" src="https://github.com/user-attachments/assets/8eada446-038a-4b05-8902-af39eb9740d9" />

## PR Content Description
This PR adds a standalone tuning script for Triton LoRA kernels, `vllm/benchmarks/kernels/benchmark_lora_tuning.py`.

**High‑level behavior**

- Performs a grid search over Triton kernel hyperparameters (e.g. `block_m/n/k`, `num_warps`, `num_stages`, `split_k`, etc.) for LoRA and fused MoE LoRA ops.
- Reuses the existing `benchmark_lora.bench_optype` harness to run the kernels and measure median latency per configuration.
- Optionally exports the best configuration(s) per setting to JSON files that are compatible with `get_lora_op_configs` (keyed by `max_loras`, `num_slices`, `m`, `k`, `n` and optionally `moe_intermediate_size` for fused MoE).

**Patching / config hijacking**

- The script introduces a small monkey‑patching layer around `get_lora_op_configs`:
  - `_CURRENT_LORA_KERNEL_CONFIG` holds the “current candidate” Triton config during tuning.
  - `_patched_get_lora_op_configs(...)` first checks `_CURRENT_LORA_KERNEL_CONFIG`:
    - If it is set, it returns this candidate config instead of running the original logic.
    - For fused MoE LoRA ops, it also maps the lower‑case search‑space keys (`block_m`, `num_warps`, `split_k`, …) to the macro‑style keys expected by the existing code (`BLOCK_SIZE_M`, `NUM_WARPS`, `SPLIT_K`, …).
    - If no candidate is set, it falls back to the original `_ORIGINAL_GET_LORA_OP_CONFIGS(...)`.
  - `_ensure_patched()` applies this patched function to all relevant modules (`utils`, `lora_shrink_op`, `lora_expand_op`, `fused_moe_lora_op`, and the `benchmark_lora` module itself) exactly once.
- `_benchmark_single_config(...)` simply sets `_CURRENT_LORA_KERNEL_CONFIG = cfg` before calling `bench_optype(...)`, so every benchmark run uses that specific kernel config via the patched `get_lora_op_configs`.
- If Triton reports `OutOfResources`, the script treats that configuration as invalid (marks status as `out_of_resources` and skips it when choosing the best config).
---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

caozuoba committed:
Improve and supplement LoRA tuning documentation

caozuoba committed:
modify

dcmaddix commented:
> Thank you for the code update and improvements! We found out if four fused_moe_lora kernels are using different `block_size_m` the illegal memory access will happen. This is because `block_size_m` is used in `moe_lora_alignment` to generate input blocks. Probably the best idea is to tune the four `fused_moe_lora` kernels together or at least make sure `block_size_m` is the same across four kernels per batch.

Yes, it is important that they use the same `block_size_m` which is also dependent on the `moe_lora_block_align` kernel. Otherwise LGTM, thanks!

caozuoba commented:
> Thank you for the code update and improvements! We found out if four fused_moe_lora kernels are using different `block_size_m` the illegal memory access will happen. This is because `block_size_m` is used in `moe_lora_alignment` to generate input blocks. Probably the best idea is to tune the four `fused_moe_lora` kernels together or at least make sure `block_size_m` is the same across four kernels per batch.

Thank you for pointing out the issue. I will make the necessary changes and update the relevant code.

caozuoba committed:
Fix illegal memory access in fused MoE LoRA via joint kernel tuning

caozuoba committed:
Fix pre-commit check issues

caozuoba committed:
Add SPDX license header

caozuoba committed:
Fix markdownlint issues

caozuoba committed:
Disable base model adapter in LoRA benchmark

caozuoba commented:
> > > Thank you for the code update and improvements! We found out if four fused_moe_lora kernels are using different `block_size_m` the illegal memory access will happen. This is because `block_size_m` is used in `moe_lora_alignment` to generate input blocks. Probably the best idea is to tune the four `fused_moe_lora` kernels together or at least make sure `block_size_m` is the same across four kernels per batch.
> > 
> > 
> > @yugong333 @dcmaddix Fixed! I've implemented joint tuning to ensure all four fused_moe_lora kernels share the same BLOCK_SIZE_M. This prevents the illegal memory access issue you mentioned. Please review when you have a chance.
> 
> Thank you very much! I also leave a small comment about `adapter_enabled` on the related code. Others LGTM! Thanks again.

@yugong333 Good point, thanks for catching that. I have fixed and submitted the new code.
